ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
       return false;
     }
-    if (!partitionId.equals(blobId.partitionId)) {
+    if (!flag.equals(blobId.flag)) {
       return false;
     }
-    if (!uuid.equals(blobId.uuid)) {
+    if (!datacenterId.equals(blobId.datacenterId)) {
       return false;
     }
-
-    return true;
+    if (!accountId.equals(blobId.accountId)) {
+      return false;
+    }
+    if (!containerId.equals(blobId.containerId)) {
+      return false;
+    }
+    if (!partitionId.equals(blobId.partitionId)) {
+      return false;
+    }
+    return uuid.equals(blobId.uuid);
   }
 
   @Override
   public int hashCode() {
     return Utils.hashcode(new Object[]{version, partitionId, uuid});
   }
+
+  /**
+   * Populate the missing fields with legacy values, and determine the version of format to serialize. Only
+   * when all the four fields of {@code flag}, {@code datacenterId}, {@code accountId}, and {@code containerId}
+   * are available, the serialization format version will be set to {@link #BLOB_ID_V2}, otherwise it will be
+   * {@link #BLOB_ID_V1}.
+   */
+  private void populateMissingFieldsAndDetermineVersion() {
+    version = flag == null && accountId == null && containerId == null && datacenterId == null ? BlobId.BLOB_ID_V1
```

Comment:
This should be same as BlobProperties. Server always deals with storeKeys. So, in first version we will ensure all servers and frontends will be able to read BlobIdV2. new BlobId() will be called only by router. So, when we start deploying 2nd version, FE_V1 will send out PUTs in BlobIdV1 and FE_V2 will send out PUTs in BlobIdV2. Since servers will be able to read both blobId versions, they should be able to handle both type of requests (from both FE_V1 and FE_V2)